### ISSUE TITLE
Depth ordering issue in menus in NBA2K13

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#16866]
Reported through email. Virtual keyboard and other popups are drawn behind the menu, when they should be above.

It reportedly broke somewhere between 1.13 and 1.14, so it's a proper regression.

![image](https://user-images.githubusercontent.com/130929/215525654-a25a971b-5afd-499c-ad01-8cd728145608.png)

GE framedump:

[NBA 2K13 ULUS10598_0001.zip](https://github.com/hrydgard/ppsspp/files/10537374/NBA.2K13.ULUS10598_0001.zip)

**Comment by Gamemulatorer** [comment]
I can only reproduce this issue on vulkan backend.

OPENGL
![Screenshot_20230131_000000_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/215528492-678e0494-da4a-40e2-867e-9a0a6c6f3abe.jpg)

VULKAN
![Screenshot_20230131_000033_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/215528569-1b3901b6-68ec-45b1-b31e-abe8fa8e3544.jpg)

**Comment by hrydgard** [comment]
I can reproduce it with the dump with OpenGL on the desktop, too. It's a depth testing problem.

It's unlikely to be related to #16530 (Ironman), that's quite a different issue (we're missing copies of depth buffers there).

**Comment by hrydgard** [comment]
Managed to do a bisect, this broke in aac2faa674edbd2a66a9332ff023883132405e9a .

**Comment by unknownbrackets** [comment]
357/659 draws the left background.  Z=2559.023438 on the draw, replacing Z=38311 with a `<=` depth test.  The interior lines are drawn using Z=1889.896484 and then Z=1771.816406, and the same depth test.  This results in 2559, 1890, and 1772.  Text uses Z=1247.011719 resulting in 1247.

At 502/659, an ALWAYS depth test is used which draws Z=65535.562500 in transform mode with depth clamp disabled.  Because this exceeds 65535 (maxz), it is not drawn.

Next, the keyboard is drawn at 503/659, with a `<=` depth test using Z=37445.968750.  Obviously, because depth was not cleared, this does not draw on top of everything as is expected.

In software rendering, which I suspect is accurate in this case, the ALWAYS full screen draw succeeds and fills Z=65535 on the entire screen.  The keyboard displays as expected.

So the reason this broke from #16383 might be a difference in how rounding is happening or we might've had a buffer around the range previously allowing this slight overflow (although that wouldn't be the case if depth clamp is supported, iirc.)

This draw uses:
 * Float positions, where Z=1000.0.
 * Transform, and projection mainly contributes to Z.
 * Viewport in a semi-standard but non-negated 32767.5 + z * 32767.5.
 * Max Z as the standard 65535 value.
 * Alpha blending to prevent any actual color changes.  Should've been clear mode.
 * 4 vertices in a triangle strip.

-[Unknown]

**Comment by hrydgard** [comment]
Thanks for looking at it!

Should comparisons against maxz round down to integer first somehow? (Could get approximately the same effect by just incrementing maxz by one...)

**Comment by unknownbrackets** [comment]
Well, we can do that - #16880.  I think on a device where the depth buffer is actually 16-bit, that won't work.  But that's pretty uncommon, I think.  It also won't work when not doing the "24-to-16" trick with depth clamp, as I explained in the comment in that pull.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, lots of details to get right and we can't trust various mobile GPUs and stuff to have exactly the same depth behavior either... But at least that should improve things, and now we have one more tricky test case for future investigations.

**Comment by hrydgard** [comment]
Reported by "Winnipeg Jets Fan" on Discord to still be happening:

![Image](https://github.com/user-attachments/assets/25c71dac-8d52-440b-9dab-9fa59ed5444e)
